inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/reduce_gpu_b_fs_yx_fsv16.cl

```diff
@@ -0,0 +1,429 @@
+    const uint batch_max_val = batch_out + 1;
+#endif
+
+#if REDUCE_FEATURE
+    const uint feature_out = 0;
+    const uint feature_max_val = INPUT0_FEATURE_NUM;
+#else
+    const uint feature_out = FEATURE_NUM_IDX_COMP(linear_idx);
+    const uint feature_max_val = feature_out + 1;
+#endif
+
+#if REDUCE_Y
+    const uint y_out = 0;
+    const uint y_max_val = INPUT0_SIZE_Y;
+#else
+    const uint y_out = SIZE_Y_IDX_COMP(linear_idx);
+    const uint y_max_val = y_out + 1;
+#endif
+
+#if REDUCE_X
+    const uint x_out = 0;
+    const uint x_max_val = INPUT0_SIZE_X / READ_OFFSET;
+    const uint x_leftover_start = x_max_val * READ_OFFSET;
+    const uint x_leftover_end = INPUT0_SIZE_X;
+#else
+    const uint x_out = SIZE_X_IDX_COMP(linear_idx);
+    const uint x_max_val = x_out + 1;
+    const uint x_leftover_start = x_out;
+    const uint x_leftover_end = x_max_val;
+#endif
+
+uint offset = batch_out * input_batch_pitch + ((feature_out + FSV - 1) / FSV) * input_fs_pitch + y_out * input_y_pitch + x_out * input_x_pitch;
+uint leftover_offset = batch_out * input_batch_pitch + ((feature_out + FSV - 1) / FSV) * input_fs_pitch + y_out * input_y_pitch + x_leftover_start * input_x_pitch;
+
+#if REDUCE_BFYX_BYX_BFX_FYX || REDUCE_BX_FX_YX_X
+    ACCUMULATOR_TYPE acc = INIT_VAL;
+    for (uint bi = batch_out; bi < batch_max_val; ++bi) {
+        for (uint fi = feature_out; fi < feature_max_val; fi += FSV) {
+            for (uint yi = y_out; yi < y_max_val; ++yi) {
+                for (uint xi = x_out; xi < x_max_val; ++xi) {
+                    INPUT_VEC input = (INPUT_VEC)(INPUT_INIT_VAL);
+                    #if (REDUCE_MAX_MODE || REDUCE_MIN_MODE || REDUCE_PROD_MODE || REDUCE_AND_MODE || REDUCE_LOG_SUM_EXP_MODE) && REDUCE_FEATURE && (INPUT0_FEATURE_NUM % FSV != 0)
+                        if (fi + get_sub_group_local_id() < INPUT0_FEATURE_NUM)
+                            for (int i = 0; i < READ_OFFSET; ++i)
+                                input[i] = data[offset + get_sub_group_local_id() + i * get_max_sub_group_size()];
+                    #else
+                        input = BLOCK_READ(data, offset);
+                    #endif
+                    unroll_for (int i = 0; i < READ_OFFSET; ++i)
+                        acc = FUNC_CALL(apply_reduce)(acc, input[i]);
+                    offset += input_x_pitch * READ_OFFSET;
+                }
+                offset += input_y_pitch - (x_max_val - x_out) * READ_OFFSET * input_x_pitch;
+            }
+            offset += input_fs_pitch - ((y_max_val - y_out) * input_y_pitch);
+        }
+        offset += input_batch_pitch - ((((feature_max_val - feature_out) + FSV - 1) / FSV) * input_fs_pitch);
+    }
+
+    for (uint bi = batch_out; bi < batch_max_val; ++bi) {
+        for (uint fi = feature_out; fi < feature_max_val; fi += FSV) { 
+            for (uint yi = y_out; yi < y_max_val; ++yi) {
+                for (uint xi = x_leftover_start; xi < x_leftover_end; ++xi) {
```

Comment:
Since those loops are needed only for x leftovers, maybe it will be more performant to have if in first loop only, similar to how feature leftovers are handled?

Reply:
Consulted it with @kdobros and decided to keep it way it is implemented now.

Reply:
Yeah, sorry I forgot to mention during our sync - three outer loops are common in both sections, so inner loop can be left, but outer loops can still be merged.